UNITTESTS/README.md

```diff
@@ -0,0 +1,341 @@
+  │       ├── NetworkInterface               Name of the class to be tested
+  │       │   ├── test_NetworkInterface.cpp
+  │       │   └── unittest.cmake             CMake module for unit test
+  │       └── Socket
+  │
+  ├── stubs                                  Shared stubs which can be used for tests.
+  ├── target_h                               Shared headers which can be used for tests.
+  └── template                               Templates for creating new unittests
+```
+Each unit test has an identical directory tree as seen in the Mbed OS root folder. This is not a mandatory requirement but helps to maintain tests. Each class to be tested have their own `unittest.cmake` which is found by `CMakeLists.txt`.
+
+## Creating a unit test
+
+Each class to be tested requires two files for unit testing:
+1. C++ unit test source file (e.g. `test_NetworkInterface.cpp`)
+2. CMake module file for unit test definition (`unittest.cmake`)
+
+A unit test definition file `unittest.cmake` requires variables to be set for a test to be configured.
+
+* **TEST_SUITE_NAME** - Identifier for the test suite. Use naming convention *PATH_TO_THE_TESTABLE_FILE* e.g. *features_netsocket_InternetSocket*
+* **unittest-includes** - Include paths for headers needed to build the tests in addition to the base include paths listed in [CMakeLists.txt](CMakeLists.txt). Optional.
+* **unittest-sources** - Mbed OS source files and stubs included for the build.
+* **unittest-test-sources** - Unit test source files.
+
+#### Creating unit tests files with Python script:
+
+```
+./UNITTESTS/mbed_unittest.py new
+```
+
+##### Parameters to mbed_unittest.py new
+
+```
+usage: mbed_unittest.py new [-h] -f file
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -f file, --filepath file
+                        specify filepath to the file to be tested
+```
+
+E.g. `./UNITTESTS/mbed_unittest.py new -f rtos/Semaphore.cpp`
+
+The generator script only creates the files required for a unit test. It does not write unit tests automatically nor does it handle source dependencies.
+
+#### Create files manually
+
+For example to create a unit test for `rtos/Semaphore.cpp`:
+
+1. Create a directory for unit test files in `UNITTESTS/rtos/Semaphore`.
+2. Create a test definition file `UNITTESTS/rtos/Semaphore/unittest.cmake` with the following content:
+```
+set(TEST_SUITE_NAME "rtos_Semaphore")
+
+set(unittest-sources
+	stubs/mbed_assert.c
+	../rtos/Semaphore.cpp
+)
+
+set(unittest-test-sources
+	rtos/Semaphore/test_Semaphore.cpp
+)
+```
+3. Create a test source file `UNITTESTS/rtos/Semaphore/test_Semaphore.cpp` with the following content:
+```
+#include  "gtest/gtest.h"
+#include  "rtos/Semaphore.h"
+
+static osStatus_t retval = osOK;
+static uint32_t count = 0;
+
+// Test stubs
+osStatus_t osSemaphoreAcquire (osSemaphoreId_t semaphore_id, uint32_t timeout) { return retval; }
```

Comment:
examples should follow our coding style (please run astyle on this code and update the example)